### PR TITLE
Add a way to parse repo/namespace/image into a DockerRegistryEndpoint

### PR DESCRIPTION
**Pull request by carlossg**
@reviewbybees

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/docker/commons/DockerRegistryEndpoint.java [hunk 1]
```diff
@@ -53,6 +72,24 @@ public DockerRegistryEndpoint(String url, String credentialsId) {
         this.credentialsId = Util.fixEmpty(credentialsId);
     }
 
+    public static DockerRegistryEndpoint parse(String repo) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/docker/commons/DockerRegistryEndpoint.java]: Certainly needs Javadoc.

**Comment by carlossg** [comment]
Added javadoc and also the missing parse of `:tag` just in case

File: src/main/java/org/jenkinsci/plugins/docker/commons/DockerRegistryEndpoint.java [hunk 2]
```diff
@@ -54,6 +71,34 @@ public DockerRegistryEndpoint(String url, String credentialsId) {
     }
 
     /**
+     * Parse the registry endpoint out of a registry:port/namespace/name:tag string. Credentials are set to the id
+     * passed.
+     * 
+     * @param s
+     * @param credentialsId
+     *            can be null
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/docker/commons/DockerRegistryEndpoint.java]: Just add <code>@CheckForNull</code> annotation to the field

File: src/main/java/org/jenkinsci/plugins/docker/commons/DockerRegistryEndpoint.java [hunk 3]
```diff
@@ -54,6 +71,34 @@ public DockerRegistryEndpoint(String url, String credentialsId) {
...
+        Matcher matcher = DOCKER_REGISTRY_PATTERN.matcher(s);
+        if (!matcher.matches() || matcher.groupCount() < 7) {
+            throw new IllegalArgumentException(s + " does not match regex " + DOCKER_REGISTRY_PATTERN);
+        }
+        String url;
+        try {
+            // docker push always uses https
+            url = matcher.group(2) == null ? null : new URL("https://" + matcher.group(2)).toString();
+        } catch (MalformedURLException e) {
+            throw new IllegalArgumentException(s + " can not be parsed as URL: " + e.getMessage());
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/docker/commons/DockerRegistryEndpoint.java]: IMO should be mentioned in Javadoc

**Comment by jglick** [comment]
You have a compilation error here.

File: src/main/java/org/jenkinsci/plugins/docker/commons/DockerRegistryEndpoint.java (hunk 2 above)
Comment (carlossg) [inline src/main/java/org/jenkinsci/plugins/docker/commons/DockerRegistryEndpoint.java]: @oleg-nenashev do you mean @Nullable in the credentialsId param?

@CheckForNull is already set in the credentialsId field

Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/docker/commons/DockerRegistryEndpoint.java]: Sorry, it was a typo. The initial message was supposed to have: "...add <code>@CheckForNull</code> annotation to the **parameter**..."

**Comment by carlossg** [comment]
damn eclipse, let's try again

**Comment by oleg-nenashev** [comment]
:+1:

**Comment by jglick** [comment]
BTW when you rebase commits inside the PR, I cannot see what you changed since the last time I looked, so I have to reread the whole thing.

File: src/test/java/org/jenkinsci/plugins/docker/commons/DockerRegistryEndpointTest.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
+package org.jenkinsci.plugins.docker.commons;
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/docker/commons/DockerRegistryEndpointTest.java]: Missing license header.

File: src/main/java/org/jenkinsci/plugins/docker/commons/DockerRegistryEndpoint.java [hunk 5]
```diff
@@ -54,6 +71,36 @@ public DockerRegistryEndpoint(String url, String credentialsId) {
     }
 
     /**
+     * Parse the registry endpoint out of a registry:port/namespace/name:tag string. Credentials are set to the id
+     * passed. The url is built from registry:port into https://registry:port, the same way docker push does.
+     * 
+     * @param s
+     * @param credentialsId
+     *            can be null
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/docker/commons/DockerRegistryEndpoint.java]: It was not obvious to me until I read the sources that this was not used as such, merely passed along to the `DockerRegistryEndpoint` constructor. Worth clarifying.

File: src/main/java/org/jenkinsci/plugins/docker/commons/DockerRegistryEndpoint.java [hunk 6]
```diff
@@ -54,6 +71,36 @@ public DockerRegistryEndpoint(String url, String credentialsId) {
...
+     * @throws IllegalArgumentException
+     *             if string can't be parsed
+     * @return The DockerRegistryEndpoint corresponding to the registry:port part of the string
+     */
+    public static DockerRegistryEndpoint parse(String s, @CheckForNull String credentialsId) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/docker/commons/DockerRegistryEndpoint.java]: `parse` is a pretty generic name, and the input string here is not really a representation of the endpoint. It is an image name, as constructed by the `imageName` method. So I would suggest calling this `fromImageName`, and adding bidirectional Javadoc links with `imageName` to make it clear that these are related operations.

File: src/test/java/org/jenkinsci/plugins/docker/commons/DockerRegistryEndpointTest.java (hunk 4 above)
Comment (carlossg) [inline src/test/java/org/jenkinsci/plugins/docker/commons/DockerRegistryEndpointTest.java]: not particularly obvious when only one file in the project has it. MIT? ASL?

Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/docker/commons/DockerRegistryEndpointTest.java]: MIT ought to be specified in the POM no?

**Comment by carlossg** [comment]
Addressed comments. Should typically rebase before merging or up to me?

File: src/test/java/org/jenkinsci/plugins/docker/commons/DockerRegistryEndpointTest.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2015 Carlos Sanchez <carlos@apache.org>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/docker/commons/DockerRegistryEndpointTest.java]: The POM does specify MIT.

**Comment by jglick** [comment]
:+1: other than incorrect license header.

Up to the merger to rebase or not. I never do.

**Comment by oleg-nenashev** [comment]
I re-confirm :+1:

**Comment by carlossg** [comment]
Merged in d0d19c6

**Pull request closed without merging**